python_modules/libraries/dagster-airlift/dagster_airlift/core/load_defs.py

```diff
@@ -371,3 +398,100 @@ def construct_dataset_specs(
             for dataset in serialized_data.datasets
         ],
     )
+
+
+def _get_dag_to_asset_mapping(
+    mapped_assets: Sequence[MappedAsset],
+) -> Mapping[str, Sequence[Union[AssetSpec, AssetsDefinition]]]:
+    res = defaultdict(list)
+    for asset in mapped_assets:
+        if is_task_mapped_asset_spec(asset):
+            for task_handle in task_handles_for_spec(asset):
+                res[task_handle.dag_id].append(asset)
+        elif is_dag_mapped_asset_spec(asset):
+            for dag_handle in dag_handles_for_spec(asset):
+                res[dag_handle.dag_id].append(asset)
+    return res
+
+
+def _dag_multi_asset(
+    dag_id: str,
+    specs: Sequence[AssetSpec],
+) -> AssetsDefinition:
+    @multi_asset(
+        specs=specs,
+        name=f"{dag_id}_underlying_op",
+        # We need to allow assets definitions to be subsettable in the case where
+        # the same asset is referenced in multiple DAGs.
+        can_subset=True,
+    )
+    def _dag_multi():
+        pass
+
+    return _dag_multi
+
+
+def build_dag_assets_defs(
+    dag_to_assets_mapping: Mapping[str, Sequence[AssetSpec]],
+) -> Sequence[AssetsDefinition]:
+    already_seen = set()
+    res = []
+    for dag_id, specs in dag_to_assets_mapping.items():
+        unseen_specs = [spec for spec in specs if spec.key not in already_seen]
+        if not unseen_specs:
+            continue
+        res.append(
+            _dag_multi_asset(
+                dag_id=dag_id,
+                specs=[spec for spec in specs if spec.key not in already_seen],
+            )
+        )
+        already_seen.update(spec.key for spec in specs)
+    return res
```

Comment:
discussed offline but should be able to just create one multi asset for all specs seen in any dag right?

might be a secondary concern if we can get the alternative external job pattern working